mantidproject/mantid
# Corelli functionality to fit a bank of tubes

**Description of work.**

- [x] function to fit a bank of tubes (and unit test)
- [x] function for the acceptance/rejection of a tube fit (and unit test)
- [x] function to assess if the number of neutron counts warrant a reliable tube fit (and unit test)
- [x] function to lo a bank selection (and unit test)
- [x] function to calibrate one bank (and unit test)
- [x] function to calibrate a bank selection (and unit test)
- [x] criterium function outputs a Workspace2D for deviation and Z-score versus tube index
- [x] function apply_calibration (can show the instrument, too)
- [x] write a test for the reviewer
- [x] document the functions' docstring

**To test:**
First make sure that the `scripts/` directory located under the Mantid source repository is added to the list of scripts directories. Then, in the script Editor of the workbench, run the following script (should take not more than 5 minutes):
```
from corelli.calibration.bank import calibrate_banks
from corelli.calibration.utils import apply_calibration, load_banks
counts_file = '/SNS/CORELLI/shared/tmp/CORELLI_124023_counts.nxs'
load_banks(counts_file, bank_selection='10-19', output_workspace='counts')
calibrate_banks('counts', bank_selection='10-19')
apply_calibration('counts', 'calib10', output_workspace='counts_10', show_instrument=True)
```

Open the instrument view for workspace `counts`, and compare `bank10` against the instrument view of workspace `counts_10`. All tubes except  tube13 have been successfully calibrated.

Within the `acceptances` WorkspaceGroup lies workspace `acceptance10`. Plot spectrum number 3 of workspace `acceptance10` showing the Z-score for each tube. Tube 13 has a large Z-score denoting a bad calibration.

We can mask tube13 using table `mask10` located within WorkspaceGroup `masks`. Run the following line:
```
MaskDetectors('counts_10', DetectorList=mtd['mask10'].column(0))
```
Now verify in the instrument view of workspace `counts_10` that its bank10  has tube13 masked.

[Release notes](https://github.com/mantidproject/mantid/blob/8b70d155b5cecf345006a14f64da6bdcf13368c3/docs/source/release/v6.0.0/diffraction.rst#new-features)

Closes #29544

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
